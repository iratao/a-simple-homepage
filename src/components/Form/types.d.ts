export interface Props {
  children: React.ReactNode
}

export interface Rule {
  required?: boolean;
  pattern?: string | RegExp;
  validator?: (value: any, form: FormInstance) => boolean;
  min?: number;
  message?: string;
}

export type Rules = Rule[]

export interface Field {
  value: string | undefined;
  error: string | undefined;
  rules: Rules | undefined;
}

export interface Fields {
  [name: string]: Field
}

export interface FormValues {
  [name: string]: string | undefined;
}

export interface FormError {
  errors: string[];
  name: string;
}

export type FormErrors = FormError[]

export type OnChangeEvent = React.ChangeEvent<HTMLInputElement>

export interface FieldOpertions {
  value?: string;
  onChange?: (event: OnChangeEvent) => void;
  onClick?: () => any;
}

export type FormItemType = undefined | 'submit'
export type RegisterFieldConfig = {
  name: string; 
  rules?: Rules;
  type?: FormItemType;
}

export interface ValidateResult {
  values: FormValues,
  errors?: FormErrors,
}

export interface FormInstance {
  registerField: (config: RegisterFieldConfig) => FieldOpertions;
  validate: () => ValidateResult;
  getFieldValue: (name: string) => any;
}