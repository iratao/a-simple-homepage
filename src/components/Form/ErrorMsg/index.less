.errorMsg {
  padding-top: 4px;
  color: red;
}