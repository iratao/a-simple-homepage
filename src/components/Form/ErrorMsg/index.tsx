import React from 'react'
import styles from './index.css'

interface ErrorMsgProps {
  error?: string
}

const ErrorMsg = ({ error }: ErrorMsgProps) => {
  if (!error) {
    return null
  }
  return <div className={styles.errorMsg}>{error}</div>
}

export default ErrorMsg;
  