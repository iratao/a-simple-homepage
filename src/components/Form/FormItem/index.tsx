import React, { useEffect, useState } from 'react'
import ErrorMsg from '../ErrorMsg'
import { FormInstance, Rules, FormItemType } from '../types'
import styles from './index.css'

interface FormItemProps {
  children: React.ReactElement;
  form?: FormInstance;
  name?: string;
  rules?: Rules;
  type?: FormItemType;
  error?: string;
  className?: string;
}

const FormItem = ({ children, name, form, rules, type, error, className }: FormItemProps) => {
  const [field, setField] = useState({})

  useEffect(() => {
    if (name && form) {
      setField({
        ...form.registerField({ name, rules, type })
      })
    }
  }, [])

  return (
    <div className={`${styles.formItem} ${className}`}>
      {React.Children.map(children, (child, index) => {
        if (child && index === 0 && form) {
          if (type === 'submit') {
            const onClick = child.props.onClick ? () => {
              child.props.onClick(form.validate())
            } : null
            return React.cloneElement(child, {
              ...field,
              ...child.props,
              onClick,
            })
          }
          if (name) {
            return React.cloneElement(child, {
              ...field,
              ...child.props,
            })
          }
        }
        return child;
      })}
      <ErrorMsg error={error} />
    </div>
  )
}

export default FormItem;