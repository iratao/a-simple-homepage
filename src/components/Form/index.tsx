import React, { useCallback, useEffect, useState, useMemo } from 'react'
import { Props, FormErrors, FormValues, FormError, Rule, Fields, RegisterFieldConfig, FormInstance, OnChangeEvent } from './types'
import ErrorMsg from './ErrorMsg/index';
import FormItem from './FormItem/index';
import { capitalize } from '../../utils'

const Form = ({ children }: Props) => {
  const [ fields, setFields ] = useState<Fields>({})

  const updateFieldError = (name: string, error: string) => {
    setFields((prevFileds) => ({
      ...prevFileds,
      [name]: {
        ...fields[name],
        error,
      }
    }))
  }

  const validate = () => {
    let errors: FormErrors | undefined = undefined
    const values: FormValues = {}
    Object.keys(fields).forEach(name => {
      values[name] = fields[name].value
      const field = fields[name]
      const { rules, value } = field
      const error: FormError = {
        name,
        errors: []
      }
      rules?.forEach((rule: Rule) => {
        const { required, validator, pattern, message, min } = rule;
        if (required) {
          if (!value) {
            const errorMsg = message || `${capitalize(name)} is required`
            updateFieldError(name, errorMsg)
            error.errors.push(errorMsg)
          }
        }
        if (pattern && value) {
          if (!RegExp(pattern).test(value)) {
            const errorMsg = message || `Invalide ${name}`
            updateFieldError(name, errorMsg)
            error.errors.push(errorMsg)
          }
        }
        if (validator) {
          if (!validator(value, form)) {
            const errorMsg = message || 'Validator failed'
            updateFieldError(name, errorMsg)
            error.errors.push(errorMsg)
          }
        }
        if (min !== undefined && value && typeof value === 'string') {
          if (value.length < min) {
            const errorMsg = message || `${capitalize(name)} should be at least ${min} characters long.`
            updateFieldError(name, errorMsg)
            error.errors.push(errorMsg)
          }
        }
      })
      if (error.errors.length > 0) {
        if (!errors) {
          errors = []
        }
        errors.push(error)
      }
    })
    return {
      values,
      errors,
    };
  }
  
  const registerField = ({ name, rules }: RegisterFieldConfig) => {
    setFields((prevFields) => ({
      ...prevFields,
      [name]: {
        value: '',
        error: '',
        rules,
      }
    }))
    return  {
      value: fields[name]?.value,
      onChange: (event: OnChangeEvent) => {
        const { target } = event
        const value = target? target.value : ''
        setFields((prevFields) => ({
          ...prevFields,
          [name]: {
            ...prevFields[name],
            value,
            error: '',
          }
        }))
      },
    }
  }

  const getFieldValue = (name: string) => {
    return fields[name]?.value
  }

  const form = useMemo(() => {
    const form: FormInstance = {
      registerField,
      validate,
      getFieldValue,
    }
    return form
  }, [ registerField ] )

  return (
    <div className={'form'}>
      {React.Children.map(children, (child) => {
        if (React.isValidElement(child)) {
          return React.cloneElement(child, { form, error: fields[child.props.name]?.error })
        }
        return child
      })} 
    </div>
  )
}


export default Form;
export {
  ErrorMsg,
  FormItem,
}