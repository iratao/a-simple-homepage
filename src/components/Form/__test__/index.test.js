import React from 'react';
import renderer from 'react-test-renderer';
import Form from '../index';
import FormItem from '../FormItem'

it('renders one form item correctly', () => {
  const tree = renderer
    .create(<Form>
      <FormItem>
        <input />
      </FormItem>
    </Form>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders form item with name property correctly', () => {
  const tree = renderer
    .create(<Form>
      <FormItem name="input">
        <input />
      </FormItem>
    </Form>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders form item of submit type correctly', () => {
  const tree = renderer
    .create(<Form >
      <FormItem type="submit">
        <button>click</button>
      </FormItem>
    </Form>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});