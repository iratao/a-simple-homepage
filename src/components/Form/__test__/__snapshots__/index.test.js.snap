// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form item of submit type correctly 1`] = `
<div
  className="form"
>
  <div
    className="formItem undefined"
  >
    <button
      onClick={null}
    >
      click
    </button>
  </div>
</div>
`;

exports[`renders form item with name property correctly 1`] = `
<div
  className="form"
>
  <div
    className="formItem undefined"
  >
    <input />
  </div>
</div>
`;

exports[`renders one form item correctly 1`] = `
<div
  className="form"
>
  <div
    className="formItem undefined"
  >
    <input />
  </div>
</div>
`;
